File: net/crates/net/tests/natsim/browser/driver/driver.mjs
// The browser-control plane for ONE namespace of the NAT conformance
// matrix.
//
// One Node process per NAT'd namespace, launched by the runner as
// `ip netns exec nsim_a node driver.mjs`. Everything below it — this
// process, Playwright, the browser, every UDP socket ICE opens — is
// therefore inside that namespace and behind that namespace's
// simulated NAT, while the NDJSON control channel stays on the stdio
// pipes the runner already holds. That is the whole reason the drivers
// are spawned per namespace instead of one driver launching two
// browsers with `executablePath` wrappers: a wrapper puts the BROWSER
// in the namespace but leaves Playwright's own sockets outside, and
// every future "why is this page reaching something it should not"
// becomes a question about the wrapper.
//
// Protocol: one JSON request per line on stdin, one JSON response per
// line on stdout, correlated by `id`; everything else goes to stderr,
// which the runner forwards into the scenario log.
//
//   {"id":1,"op":"launch","engine":"chromium","spkiPin":"…","caPem":"…","profileDir":"…"}
//   {"id":2,"op":"open","url":"http://localhost:8080/?tab=a&control=…"}
//   {"id":3,"op":"shutdown"}
//
// TLS: there is no `ignoreHTTPSErrors` here and there cannot be. The
// anchor's listener serves a leaf issued by the run's own CA, and each
// engine is told about that key the narrowest prompt-free way it
// offers — Chromium a one-key SPKI pin, Firefox the CA in the NSS
// database inside its own launch profile. Verification stays on.

import { chromium, firefox } from 'playwright-core';
import { execFileSync, spawn } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';
import os from 'node:os';
import readline from 'node:readline';

const ENGINES = { chromium, firefox };
const NSS_NICKNAME = 'net-mesh-natsim-harness-ca';

let live = null;
let page = null;
let internals = null;
let wildcardPorts = 0;
let realNetworkPorts = 0;
const networkNames = new Set();

function log(line) {
  process.stderr.write(line + '\n');
}

function reply(obj) {
  process.stdout.write(JSON.stringify(obj) + '\n');
}

// Firefox reads its OWN NSS database (`cert9.db`) from inside the
// profile, so the profile must be seeded BEFORE launch and the launch
// must be persistent. `certutil` is the same NSS tool the Stage 4b
// harness uses on Linux; a run that cannot seed the profile is
// REFUSED, never weakened — an unverified TLS leg would prove
// something no deployment can rely on.
function seedFirefoxProfile(profileDir, caPemPath) {
  fs.mkdirSync(profileDir, { recursive: true });
  execFileSync('certutil', [
    '-d',
    `sql:${profileDir}`,
    '-A',
    '-t',
    'C,,',
    '-n',
    NSS_NICKNAME,
    '-i',
    caPemPath,
  ]);
  return `certutil -d sql:${profileDir}`;
}

function chromiumArgs(spkiPin) {
  return [
    // The netns rows run as root (netns + nft need it), and Chromium
    // refuses its sandbox as root. This is a throwaway browser in a
    // throwaway namespace; the alternative is no row at all.
    '--no-sandbox',
    '--disable-dev-shm-usage',
    // One key, this process, this run. NOT
    // `--ignore-certificate-errors`: every other certificate is still
    // verified.
    `--ignore-certificate-errors-spki-list=${spkiPin}`,
    // mDNS host-candidate obfuscation is left ON — production cannot
    // turn it off, and the two browsers sit on DIFFERENT private
    // subnets here, so a host candidate could not have connected them
    // anyway. Every direct pair in this matrix is reached through the
    // anchor's STUN (server-reflexive) or peer-reflexively, which is
    // exactly what the NAT flavors are being tested against.
    // The engine's own account, into stderr WE own. Two earlier
    // attempts failed and the reason is worth keeping: these rows used
    // to launch Playwright's headless-SHELL build, which ignores
    // `--vmodule`, and Playwright's own `launch()` swallows the
    // stderr; `--log-file` was the wrong target too, because child
    // processes inherit stderr, not the parent's log file, and ICE
    // runs in a child.
    '--enable-logging=stderr',
    // `--v=1` and three modules, down from eighteen.
    //
    // Every line the [chromium-ice] print keeps is either `INFO`
    // (libwebrtc's own `RTC_LOG`, emitted as soon as logging is on:
    // `basic_port_allocator.cc`'s `Net[…]`, `Count of networks`,
    // `Allocate ports on`, `port.cc`'s network cost, the gathered
    // candidates) or `VERBOSE1` (`filtering_network_manager.cc`'s
    // permission status, `ipc_network_manager.cc`'s network list,
    // `socket_udp.cc`'s local address) — read off the severity tokens
    // of run 35138824048's log. What the removed fifteen modules added
    // was hundreds of `connection.cc` ping lines per second, which is
    // what pushed the job-log print past its own limit.
    '--v=1',
    // Three entries, each named for what it feeds:
    //
    // `basic_port_allocator` and `port` are the PRECONDITION's own
    // source — the `Net[…]` descriptors this driver counts and
    // `run_row` refuses a Chromium row without. They are INFO lines
    // and `--v=1` alone should carry them; they stay named because a
    // precondition that silently loses its own input would fail a
    // working row, which is worse than a wide flag.
    //
    // `peer_connection_dependency_factory` is a RECORD rather than a
    // diagnosis: it logs `WebRTC routing preferences: policy: …
    // multiple_routes: …` at VLOG(3) — the renderer's own statement of
    // the EFFECTIVE `WebRtcIPHandling` value, at the point where it
    // decides whether to enumerate interfaces at all. `chrome://policy`
    // would only show a *configured* enterprise policy, and renders it
    // inside a shadow root; this is the value that actually governs,
    // from the line that reads it. (The neighbouring `Active
    // WebRtcIPHandlingPolicy` line is a `DVLOG`, compiled out of a
    // release build, which is why the level is 3 and not 1.)
    '--vmodule=basic_port_allocator=2,port=2,peer_connection_dependency_factory=3',
    // mDNS obfuscation stays ON. Turning it off was a one-cycle
    // experiment and it EXONERATED mDNS: with a real host address
    // Chromium failed identically (`sent=192 gotResponse=0`), so the
    // switch is gone rather than left behind as a knob nobody should
    // reach for.
  ];
}

async function opLaunch(req) {
  const engine = ENGINES[req.engine];
  if (!engine) throw new Error(`unknown engine: ${req.engine}`);
  let trust;
  if (req.engine === 'firefox') {
    const caPemPath = path.join(req.profileDir, 'harness-ca.pem');
    fs.mkdirSync(req.profileDir, { recursive: true });
    fs.writeFileSync(caPemPath, req.caPem);
    trust = seedFirefoxProfile(req.profileDir, caPemPath);
    const context = await firefox.launchPersistentContext(req.profileDir, {
      headless: true,
    });
    live = { engine: req.engine, context, browser: null, persistent: true };
  } else {
    // SPAWNED BY US, then attached over CDP.
    //
    // `--log-file` does not reach the renderer: Chromium's child
    // processes inherit stderr, not the parent's log file, and ICE
    // runs in the renderer. Thirteen thousand lines of
    // `chrome_debug.log` from one run were all the BROWSER process,
    // with nothing from `connection`, `stun_request` or
    // `p2p_transport_channel` even after the `--vmodule` names were
    // corrected. Playwright's own `launch()` swallows that stderr, so
    // the only way to read it is to own the process: spawn the
    // executable, keep its stderr, and attach with
    // `connectOverCDP`.
    //
    // The browser still runs in THIS namespace, which is the whole
    // point of a driver per namespace — this replaces how the process
    // is started, not where.
    fs.mkdirSync(req.profileDir, { recursive: true });
    const headed = !!process.env.DISPLAY;
    const exe = chromium.executablePath();
    const args = [
      ...chromiumArgs(req.spkiPin),
      '--remote-debugging-port=0',
      // OUTSIDE the uploaded tree. This profile contains unix
      // sockets, and `upload-artifact` refuses a tree containing
      // them — two runs uploaded NOTHING because of it, so the
      // evidence a cycle existed for was unreadable for a reason
      // unrelated to the measurement.
      `--user-data-dir=${fs.mkdtempSync(path.join(os.tmpdir(), 'natsim-cdp-'))}`,
      ...(headed ? [] : ['--headless=new']),
      'about:blank',
    ];
    const child = spawn(exe, args, { stdio: ['ignore', 'pipe', 'pipe'] });
    let wsEndpoint = null;
    const ready = new Promise((resolve, reject) => {
      const timer = setTimeout(
        () => reject(new Error('no DevTools endpoint within 60s')),
        60_000,
      );
      let buffered = '';
      child.stderr.on('data', (chunk) => {
        const text = String(chunk);
        if (!wsEndpoint) {
          buffered += text;
          const m = buffered.match(/DevTools listening on (ws:\/\/\S+)/);
          if (m) {
            wsEndpoint = m[1];
            clearTimeout(timer);
            resolve(wsEndpoint);
          }
        }
        // Only the enumeration and ICE accounting, not thirteen
        // thousand lines of dbus and histograms: the log is evidence,
        // and evidence nobody can find is not evidence.
        for (const line of text.split('\n')) {
          if (
            /basic_port_allocator\.cc|port\.cc|p2p_transport_channel\.cc|filtering_network_manager\.cc|ipc_network_manager\.cc|peer_connection_dependency_factory\.cc|socket_udp\.cc/.test(
              line,
            )
          ) {
            log(`[chromium-ice] ${line.trim()}`);
          }
          // PER LINE, not per chunk: the first version tested the whole
          // stderr chunk, so it reported real=0 wildcard=0 on a run
          // whose log was full of `Net[…Wildcard…]` — a counter that
          // measured its own buffering rather than the browser.
          if (/Net\[/.test(line)) {
            if (/Wildcard/.test(line)) wildcardPorts += 1;
            else realNetworkPorts += 1;
            // The descriptor itself, deduplicated, so the row's
            // precondition failure can name the interface the engine
            // did or did not enumerate instead of only counting.
            const m = /Net\[([^\]]*)\]/.exec(line);
            if (m) networkNames.add(m[1]);
          }
        }
        // THE PRECONDITION, pinned so it cannot regress silently.
        //
        // Ports on `Net[any:0.0.0.x/0:Wildcard:id=0]` at cost 999 mean
        // ZERO enumerated interfaces and a wildcard bind, which drops
        // every inbound datagram before STUN parsing. A real network
        // (`Net[eth0:192.168.10x.x/24:Ethernet:id=1]`) is what makes a
        // Chromium row able to work at all, and `run_row` REFUSES a
        // Chromium tab that reports none.
      });
      child.on('error', reject);
      child.on('exit', (code) =>
        reject(new Error(`chromium exited with ${code} before DevTools`)),
      );
    });
    await ready;
    log(`chromium ${headed ? 'HEADED' : 'headless'} pid ${child.pid}, cdp ${wsEndpoint}`);
    const browser = await chromium.connectOverCDP(wsEndpoint);
    const context = browser.contexts()[0] || (await browser.newContext());
    // Opened NOW, not at shutdown. `chrome://webrtc-internals` only
    // records peer connections that exist while it is open, and the
    // leaf closes its connection the moment `connect` gives up.
    try {
      internals = await context.newPage();
      await internals.goto('chrome://webrtc-internals', {
        waitUntil: 'domcontentloaded',
        timeout: 15_000,
      });
    } catch (e) {
      internals = null;
      log(`webrtc-internals open: ${e && e.message}`);
    }
    live = { engine: req.engine, context, browser, persistent: false, child };
    trust = 'spki-pin';
  }
  return { engine: req.engine, trust };
}

async function opOpen(req) {
  if (!live) throw new Error('open before launch');
  // THE §6.12 FIX, and it is a permission, not a network — BUT IT IS
  // NOW THE ROW'S CHOICE, NOT THIS DRIVER'S.
  //
  // Chromium gates local-interface enumeration on media permission:
  // `FilteringNetworkManager` logged `received permission status:
  // denied` on every failing row, so the network list the browser
  // process had ALREADY delivered was withheld from WebRTC, which
  // then allocated `any any` wildcard ports — and a wildcard port
  // drops every inbound packet before STUN parsing. That is why
  // Chromium answered no Binding Request and credited no response
  // while every packet on the wire was valid, correctly
  // credentialed and correctly addressed, and why Firefox, which
  // has no such gate, was fine throughout.
  //
  // Granting camera/microphone for the page's own origin is what a
  // real user does before a call, and it is what the six conformance
  // rows and the Firefox control do. It is a HARNESS grant: no
  // product behaviour changes, no candidate is relabelled, no
  // deadline widened.
  //
  // `req.media === 'none'` is the PERMISSION-FREE leg. "The product
  // calls no media API" is true and is source evidence about the
  // product; it is not a measurement of the browsing context that
  // was never asked, and only a leg that withholds the grant can
  // make one. So the grant is a per-request decision and this driver
  // reports back what it actually did — an unreported grant and no
  // grant must not look the same to the runner.
  let media = 'none';
  if (req.media !== 'none' && live.engine !== 'firefox') {
    const origin = new URL(req.url).origin;
    // NOT swallowed. A failed `grantPermissions` used to be logged
    // and the row continued as though the grant had happened, which
    // is the same collapse as an unreported grant: the row would then
    // measure the ungranted environment while its verdict said
    // `granted`.
    await live.context.grantPermissions(['camera', 'microphone'], { origin });
    media = 'granted';
    log(`granted camera+microphone for ${origin}`);
  } else if (req.media === 'none') {
    log('media permission WITHHELD: product defaults, nothing granted, nothing prompted');
  } else {
    // Firefox has no such gate and Playwright cannot grant camera or
    // microphone to it at all, so the control row has always run
    // ungranted. Reported as the fact it is instead of as the flag
    // the runner passed.
    log('firefox: no media permission granted (no such gate, and no Playwright support)');
  }
  page = await live.context.newPage();
  page.on('console', (m) => log(`[page] ${m.type()}: ${m.text()}`));
  page.on('pageerror', (e) => log(`[page] ERROR ${e && e.message}`));
  // `domcontentloaded`, not `load`: the page's own work starts on
  // module evaluation and the runner drives it over HTTP from there.
  // Waiting for `load` would block on whatever the page is already
  // doing.
  await page.goto(req.url, { waitUntil: 'domcontentloaded', timeout: 60_000 });
  return { url: page.url(), media };
}

// What the port allocator enumerated, as a value the runner can
// assert on rather than a line a human has to find.
function networksSnapshot() {
  return { real: realNetworkPorts, wildcard: wildcardPorts, nets: [...networkNames] };
}

function reportNetworkEnumeration(at) {
  // Named plainly, both ways: the enumerated interface is a
  // precondition of every GRANTED Chromium row, and a run that
  // silently lost it would otherwise look like a new mystery.
  //
  // The line no longer claims inbound packets are dropped. §11.8
  // measured a `real=0` pair reaching the STUN endpoint, gathering
  // srflx and delivering application payloads both ways, so that
  // clause was a conclusion this instrument cannot support. It
  // reports what it counted, and names the permission gate that
  // explains a zero.
  const nets = networkNames.size === 0 ? '(none)' : [...networkNames].join(' ');
  log(
    `[chromium-ice] PRECONDITION networks at ${at}: real=${realNetworkPorts} ` +
      `wildcard=${wildcardPorts} nets=${nets}` +
      (realNetworkPorts === 0
        ? ' — ZERO enumerated networks; expected when no media permission was granted' +
          ' (FilteringNetworkManager gates the network list), REQUIRED of a granted row' +
          ' (S6_REPORT.md §6.12, §11.8)'
        : ''),
  );
}

// `chrome://webrtc-internals`, read before the browser closes.
//
// It is the engine's own record of every peer connection in this
// process: the ICE event log, the candidate pairs and their state
// transitions, and the getStats history — the account of a failing
// check that no external observer could supply. Chromium only, opened
// in its own tab so the row's page is untouched, and best-effort: a
// dump that fails must not fail a row.
async function dumpWebrtcInternals() {
  if (!internals) return;
  try {
    // A beat for the page to render the last events it received.
    await internals.waitForTimeout(1_500);
    const text = await internals.evaluate(() => document.body.innerText || '');
    for (const line of text.split('\n')) {
      const trimmed = line.trim();
      if (trimmed) log(`[webrtc-internals] ${trimmed}`);
    }
  } catch (e) {
    log(`webrtc-internals: ${e && e.message}`);
  }
}

async function opShutdown() {
  reportNetworkEnumeration('shutdown');
  await dumpWebrtcInternals();
  try {
    if (live && live.persistent) await live.context.close();
    else if (live && live.browser) await live.browser.close();
    // `connectOverCDP` detaches rather than terminating, so the
    // process we spawned is ours to end.
    if (live && live.child && live.child.exitCode === null) live.child.kill('SIGTERM');
  } catch (e) {
    log(`shutdown: ${e && e.message}`);
  }
  live = null;
  page = null;
  internals = null;
  // The counters travel in the reply: `run_row` refuses a Chromium
  // row that enumerated no interface, and it can only do that if the
  // number reaches it.
  return { networks: networksSnapshot() };
}

const OPS = { launch: opLaunch, open: opOpen, shutdown: opShutdown };

// Requests are served strictly in order: the runner issues one at a
// time, and an `open` that raced a `launch` would report a page on no
// browser at all.
// At EXIT as well as at shutdown: a driver whose row relaunched the
// browser prints a shutdown line before the engine has logged
// anything, and that zero says nothing about the run. The exit line
// is the one that covers the whole process.
process.on('exit', () => reportNetworkEnumeration('exit'));

const rl = readline.createInterface({ input: process.stdin });
let queue = Promise.resolve();

rl.on('line', (line) => {
  if (!line.trim()) return;
  let req;
  try {
    req = JSON.parse(line);
  } catch (e) {
    log(`unparseable request: ${line}`);
    return;
  }
  queue = queue.then(async () => {
    const op = OPS[req.op];
    if (!op) {
      reply({ id: req.id, ok: false, error: `unknown op: ${req.op}` });
      return;
    }
    try {
      const out = await op(req);
      reply({ id: req.id, ok: true, ...out });
    } catch (e) {
      reply({ id: req.id, ok: false, error: (e && (e.message || String(e))) || 'unknown' });
    }
  });
});

rl.on('close', async () => {
  await queue;
  await opShutdown();
  process.exit(0);
});

// The namespace this driver — and therefore its browser, and every
// UDP socket ICE opens — is actually in, stated as a FACT observed
// from inside it rather than as the name the runner passed. The whole
// premise of the matrix is that tab a sits behind gateway a and tab b
// behind gateway b; `192.168.101.x` vs `192.168.102.x` is what makes
// that checkable when a row fails asymmetrically.
const addrs = Object.entries(os.networkInterfaces())
  .flatMap(([dev, list]) =>
    (list || []).filter((a) => a.family === 'IPv4').map((a) => `${dev}=${a.address}`),
  )
  .join(' ');
log(`netns ${process.env.NATSIM_NETNS || '(unnamed)'} addrs ${addrs} home ${process.env.HOME}`);

// The default-local-address probe, from inside the namespace.
//
// libwebrtc (and Chromium's network service, which is what actually
// enumerates here) learns a "default local address" by connect()ing a
// throwaway UDP socket at a public IP; a namespace where that route
// does not resolve enumerates nothing. `setup.sh` already gives every
// browser namespace `default via <gateway>`, so it resolves — and the
// point of printing it is that the precondition is then a FACT in the
// log rather than an inference from the topology script. Reachability
// is irrelevant: nothing answers 8.8.8.8 in this lab and nothing has
// to.
try {
  const route = execFileSync('ip', ['route', 'get', '8.8.8.8'], { encoding: 'utf8' });
  log(`ip route get 8.8.8.8 -> ${route.trim().replace(/\s+/g, ' ')}`);
} catch (e) {
  log(`ip route get 8.8.8.8 FAILED: ${(e && e.message) || e}`);
}

// Is a `WebRtcIPHandling` policy being inherited from disk?
//
// The suspicion was that the headed CDP launch picks up a policy file
// that turns interface enumeration off. Recorded once, from the only
// places Chromium reads managed policy on Linux; the EFFECTIVE value
// is reported by the renderer itself (`peer_connection_dependency_
// factory.cc`: `WebRTC routing preferences: policy: …`) in the
// [chromium-ice] lines.
const policyDirs = [
  '/etc/chromium/policies/managed',
  '/etc/chromium/policies/recommended',
  '/etc/opt/chrome/policies/managed',
  '/etc/opt/chrome/policies/recommended',
];
log(
  `chromium managed-policy files: ${
    policyDirs
      .flatMap((dir) => {
        try {
          return fs.readdirSync(dir).map((f) => `${dir}/${f}`);
        } catch {
          return [];
        }
      })
      .join(' ') || '(none)'
  }`,
);

reply({ id: 0, ok: true, hello: 'natsim browser driver', engines: Object.keys(ENGINES) });

File: net/crates/net/tests/natsim/browser/src/driver.rs
//! The runner's handle on one namespace's browser driver
//! (`driver/driver.mjs`).
//!
//! One `node` child per NAT'd namespace, launched through
//! `ip netns exec`, NDJSON on its stdio, one request at a time,
//! correlated by id. Everything the driver prints on stderr is
//! forwarded with the namespace's label so a page console line and a
//! harness line land in the same scenario log.
//!
//! This module is the only place the runner knows what a browser
//! process is, and `ip netns exec` is the only line that knows where
//! it runs.

use std::path::Path;
use std::process::Stdio;
use std::time::Duration;

use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader, Lines};
use tokio::process::{Child, ChildStdin, ChildStdout, Command};

/// How long any single driver request may take. `open` navigates a
/// real page in a real browser inside a NAT'd namespace; the rest are
/// milliseconds.
const REQUEST_TIMEOUT: Duration = Duration::from_secs(120);

pub struct Driver {
    label: String,
    child: Child,
    stdin: ChildStdin,
    lines: Lines<BufReader<ChildStdout>>,
    next_id: u64,
}

/// What the engine's own port allocator did about interface
/// enumeration, as the driver counted it over the browser's whole
/// lifetime.
///
/// The §6.12 signature is `real == 0 && wildcard > 0`: zero
/// enumerated interfaces, ports bound to the `any` address, and every
/// inbound datagram dropped before STUN parsing. `run_row` refuses a
/// Chromium row that reports it.
#[derive(Debug, Default, Clone)]
pub struct Networks {
    /// Ports allocated on a real, named network (`Net[eth0:…]`).
    pub real: u64,
    /// Ports allocated on the wildcard `any` network.
    pub wildcard: u64,
    /// The network descriptors seen, verbatim and deduplicated.
    pub nets: Vec<String>,
}

impl Driver {
    /// Launch the driver INSIDE `netns`.
    ///
    /// `ip netns exec` from this process (which is itself inside
    /// `nsim_wan`) works because `setns` is a capability question, not
    /// a topology one, and the script already runs as root. The child
    /// keeps the stdio pipes created here, so the control channel
    /// crosses the namespace boundary while every socket the driver,
    /// Playwright and the browser open does not.
    ///
    /// `home` and `runtime` are a ROOT-owned pair, and Firefox needs
    /// them. The rows run as root (netns + nft), and Firefox refuses
    /// to start at all when it is root while `$XDG_RUNTIME_DIR`
    /// belongs to somebody else — on the GitHub runner that is
    /// `/run/user/1001`, owned by `runner`, inherited straight
    /// through `sudo`. It exits 1 with
    /// `Running Nightly as root in a regular user's session is not
    /// supported.`, which Playwright surfaces as the far less
    /// informative `Target page, context or browser has been closed`.
    /// Chromium does not care, so this is set unconditionally rather
    /// than per engine: one environment for both engines is one thing
    /// to reason about.
    pub async fn spawn(
        netns: &str,
        driver_dir: &Path,
        label: &str,
        home: &Path,
        runtime: &Path,
    ) -> Result<Self, String> {
        let mut child = Command::new("ip")
            .args(["netns", "exec", netns, "node", "driver.mjs"])
            .current_dir(driver_dir)
            .env("HOME", home)
            .env("XDG_RUNTIME_DIR", runtime)
            .env("NATSIM_NETNS", netns)
            .stdin(Stdio::piped())
            .stdout(Stdio::piped())
            .stderr(Stdio::piped())
            .kill_on_drop(true)
            .spawn()
            .map_err(|e| format!("spawn driver in {netns}: {e} (is iproute2 installed?)"))?;
        let stdin = child.stdin.take().ok_or("driver stdin")?;
        let stdout = child.stdout.take().ok_or("driver stdout")?;
        let stderr = child.stderr.take().ok_or("driver stderr")?;
        let tag = label.to_owned();
        tokio::spawn(async move {
            let mut lines = BufReader::new(stderr).lines();
            while let Ok(Some(line)) = lines.next_line().await {
                println!("[{tag}] {line}");
            }
        });
        let mut driver = Self {
            label: label.to_owned(),
            child,
            stdin,
            lines: BufReader::new(stdout).lines(),
            next_id: 1,
        };
        // The hello line, so "the driver never started" is reported
        // here rather than as a timeout on the first real request.
        let hello = driver.read_line().await?;
        if hello.get("hello").is_none() {
            return Err(format!(
                "{label}: driver did not greet: {hello} (npm ci / playwright install?)"
            ));
        }
        Ok(driver)
    }

    async fn read_line(&mut self) -> Result<serde_json::Value, String> {
        let line = tokio::time::timeout(REQUEST_TIMEOUT, self.lines.next_line())
            .await
            .map_err(|_| format!("{}: driver timed out", self.label))?
            .map_err(|e| format!("{}: driver stdout: {e}", self.label))?
            .ok_or_else(|| format!("{}: driver exited", self.label))?;
        serde_json::from_str(&line)
            .map_err(|e| format!("{}: unparseable driver line {line:?}: {e}", self.label))
    }

    async fn request(
        &mut self,
        op: &str,
        mut req: serde_json::Value,
    ) -> Result<serde_json::Value, String> {
        let id = self.next_id;
        self.next_id += 1;
        req["id"] = serde_json::json!(id);
        req["op"] = serde_json::json!(op);
        let mut buf = serde_json::to_vec(&req).map_err(|e| e.to_string())?;
        buf.push(b'\n');
        self.stdin
            .write_all(&buf)
            .await
            .map_err(|e| format!("{}: driver stdin: {e}", self.label))?;
        self.stdin
            .flush()
            .await
            .map_err(|e| format!("{}: driver flush: {e}", self.label))?;
        let reply = self.read_line().await?;
        if reply.get("id").and_then(serde_json::Value::as_u64) != Some(id) {
            return Err(format!(
                "{}: driver replied out of order: {reply}",
                self.label
            ));
        }
        if reply.get("ok").and_then(serde_json::Value::as_bool) != Some(true) {
            return Err(format!(
                "{}: {op} failed: {}",
                self.label,
                reply
                    .get("error")
                    .and_then(serde_json::Value::as_str)
                    .unwrap_or("(no error field)")
            ));
        }
        Ok(reply)
    }

    /// Start the engine. `spki_pin` is Chromium's one-key pin;
    /// `ca_pem` is what Firefox's profile NSS database is seeded with.
    /// Which one an engine uses is the driver's business.
    pub async fn launch(
        &mut self,
        engine: &str,
        spki_pin: &str,
        ca_pem: &str,
        profile_dir: &Path,
    ) -> Result<String, String> {
        let reply = self
            .request(
                "launch",
                serde_json::json!({
                    "engine": engine,
                    "spkiPin": spki_pin,
                    "caPem": ca_pem,
                    "profileDir": profile_dir.to_string_lossy(),
                }),
            )
            .await?;
        Ok(reply
            .get("trust")
            .and_then(serde_json::Value::as_str)
            .unwrap_or("(unreported)")
            .to_owned())
    }

    /// Open the row's page, and report what the driver GRANTED it.
    ///
    /// `media` is `granted` or `none` and is the row's own field, not
    /// a runner preference: Chromium withholds interface enumeration
    /// from WebRTC until a media permission exists (§6.12), so the
    /// grant changes the networking environment the product runs in
    /// and a row that acquired it silently would be measuring a
    /// different environment under its own name.
    ///
    /// The returned string is what the DRIVER says it did, read back
    /// rather than echoed from the request, so a driver that ignored
    /// `media` fails the row instead of passing under the label the
    /// runner asked for.
    pub async fn open(&mut self, url: &str, media: &str) -> Result<String, String> {
        let reply = self
            .request("open", serde_json::json!({ "url": url, "media": media }))
            .await?;
        Ok(reply
            .get("media")
            .and_then(serde_json::Value::as_str)
            .ok_or_else(|| {
                format!(
                    "{}: the driver did not report what it granted the page — an unreported \
                     grant is indistinguishable from none",
                    self.label
                )
            })?
            .to_owned())
    }

    /// Stop the engine, and report what it enumerated.
    ///
    /// The counters come back in the `shutdown` reply rather than
    /// being parsed out of the log by a second reader: the driver is
    /// already counting them line by line, and a number the runner
    /// can assert on is worth more than one a human can find.
    pub async fn shutdown(mut self) -> Networks {
        let reply = self.request("shutdown", serde_json::json!({})).await;
        drop(self.stdin);
        let _ = tokio::time::timeout(Duration::from_secs(20), self.child.wait()).await;
        let Ok(reply) = reply else {
            return Networks::default();
        };
        let networks = &reply["networks"];
        Networks {
            real: networks["real"].as_u64().unwrap_or(0),
            wildcard: networks["wildcard"].as_u64().unwrap_or(0),
            nets: networks["nets"]
                .as_array()
                .map(|v| {
                    v.iter()
                        .filter_map(|n| n.as_str().map(str::to_owned))
                        .collect()
                })
                .unwrap_or_default(),
        }
    }
}
